doc: ACRN Inter-VM Communication (#6973)
projectacrn/acrn-hypervisor: doc/tutorials/inter-vm_communication.rst

.. _inter-vm_communication:

ACRN Inter-VM Communication
##############################################

ACRN supports three kinds of Inter-VM communication – Inter-VM vUART,
Inter-VM network communication, and Inter-VM shared memory device (ivshmem).
Each communication has its advantage and shortage and is listed below:

Inter-VM vUART
**************

Inter-VM vUART communication is based on the vUART implementation. It is
used to transfer data between two VMs at low speed. (Refer to :ref:`vuart_virtualization`
and :ref:`vuart_config`).

:numref:`Inter-VM vUART communication` shows the Inter-VM vUART communication overview:

.. figure:: images/Inter-VM_vUART_communocation_overview.png
   :align: center
   :name: Inter-VM vUART communication

   Inter-VM vUART communication

- Pros:
   - POSIX APIs.

>    - UART driver is available in most VM, development-friendly.

Reviewer: What do we mean by "development-friendly"?

Author: Fixed. POSIX can be called directly to realize vUART communication without additional development drivers

- Cons:
   - The communication rate is low.
   - Need to configure communication vUART in hypervisor scenario.

Inter-VM network communication
******************************

Inter-VM network communication is based on the network stack. ACRN supports
both pass-through NICs to VM and Virtio-Net solutions. (Refer to :ref:`virtio-net`
background introductions of ACRN Virtio-Net Architecture and Design). 

:numref:`Inter-VM network communication` shows the Inter-VM network communication overview:

.. figure:: images/Inter-VM_network_communication_overview.png
   :align: center
   :name: Inter-VM network communication

   Inter-VM network communication

- Pros:
   - Socket-based APIs; development friendly.

- Cons:
   - Multiple layers are involved across the data path.

>    - Potential more CPU overhead in SOS if using Virtio-net.

Reviewer: ```suggestion
   - Potential more CPU overhead in the Service VM if using ``virtio-net``.
```

Author: fixed.

Inter-VM shared memory communication (ivshmem)
**********************************************

Inter-VM shared memory communication is based on a shared memory mechanism
to transfer the data between VMs. The ACRN device model or hypervisor emulates
a virtual PCI device (called an ivshmem device) to expose this shared memory's
base address and size. (Refer to :ref:`ivshmem-hld` and :ref:`enable_ivshmem` for the
background introductions).

:numref:`Inter-VM shared memory communication` shows the Inter-VM shared memory communication overview:

.. figure:: images/Inter-VM_shared_memory_communication_overview.png
   :align: center
   :name: Inter-VM shared memory communication

   Inter-VM shared memory communication

- Pros:
   - Shared memory is exposed to VMs via PCI MMIO Bar and is mapped and accessed directly.
   - The data path is short.

>    - No SOS emulation is involved if using hv-land ivshmem.

Reviewer: ```suggestion
   - No Service VM emulation is involved if using hv-land ivshmem.
```

Author: fixed.

- Cons:

>    - Applications talk to the device directly (via UIO on Linux and Ivshmem driver on WAAG).

Reviewer: ```suggestion
   - Applications talk to the device directly (via UIO on Linux and Ivshmem driver on Windows).
```

Author: fixed.

   - Applications need to implement protocols such as a handshake, data transfer, and data
     integrity.

How to implement an Ivshmem application on ACRN
***********************************************

As Ivshmem is a PCI device that shares memory inside a virtual machine, the application
running on the VM can access it as a standard PCI device. Besides that, the application
needs to design a data transfer notification mechanism between the VMs.

- Initial Device:

>   Enable and bind to UIO PCI driver

Reviewer: Please turn the sequence below into a `code-block`

Author: fixed.

Author: Thanks, `sudo` has been added.

     - modprobe uio
     - modprobe uio_pci_generic
     - echo "1af4 1110" > /sys/bus/pci/drivers/uio_pci_generic/new_id

  .. note:: "1af4 "is the Vendor ID and "1110" is the Device ID of the ivshmem device.

  .. note:: For Linux-based User VMs, we recommend using the standard UIO and UIO_PCI_GENERIC

>    drivers through the device node (for example, /dev/uioX)

Reviewer: ```suggestion
   drivers through the device node (for example, ``/dev/uioX``)
```

Author: fixed.

.. _inter-vm_communication:

ACRN Inter-VM Communication
##############################################

ACRN supports three kinds of Inter-VM communication – Inter-VM vUART,
Inter-VM network communication, and Inter-VM shared memory device (ivshmem).

> Each communication has its advantage and shortage and is listed below:

Author: Thanks, has been fixed.

Inter-VM vUART
**************

Inter-VM vUART communication is based on the vUART implementation. It is

> used to transfer data between two VMs at low speed. (Refer to :ref:`vuart_virtualization`

Author: Thanks, has been fixed.

and :ref:`vuart_config`).

:numref:`Inter-VM vUART communication` shows the Inter-VM vUART communication overview:

.. figure:: images/Inter-VM_vUART_communocation_overview.png
   :align: center
   :name: Inter-VM vUART communication

   Inter-VM vUART communication

- Pros:
   - POSIX APIs; development-friendly (easily used programmatically
     by using standard POSIX APIs).

>    - UART driver is available in most VM.

Author: Thanks, has been fixed.

- Cons:
   - The communication rate is low.

>    - Need to configure communication vUART in hypervisor scenario.

Author: yes. Thanks, has been fixed.

Inter-VM network communication
******************************

Inter-VM network communication is based on the network stack. ACRN supports

> both pass-through NICs to VM and Virtio-Net solutions. (Refer to :ref:`virtio-net`

Author: Thanks, has been fixed.

background introductions of ACRN Virtio-Net Architecture and Design). 

:numref:`Inter-VM network communication` shows the Inter-VM network communication overview:

.. figure:: images/Inter-VM_network_communication_overview.png
   :align: center
   :name: Inter-VM network communication

   Inter-VM network communication

- Pros:

>    - Socket-based APIs; development-friendly (easily used programmatically

Author: This communication bandwidth depends on the speed of the network card.

>    - Socket-based APIs; development-friendly (easily used programmatically

Reviewer: One could also create a virtual switch/bridge and use that to establish a communication path between two (or more) VMs. Is that correct?

I think it's safe to say that even without knowing the network card HW or implementation, that communication mechanism is orders of magnitude faster than a vUART, isn't it?

Another benefit here is that this is not limited to a 1-1 communication between two VMs, more can be involved.

>    - Socket-based APIs; development-friendly (easily used programmatically

Author: Orders of magnitude faster than vUART, but can't quantify this with BPS numbers.

     by using Socket-based APIs).

- Cons:

>    - Multiple layers are involved across the data path.

Reviewer 2: Tell us what's the specific consequences are of "multiple layers" being involved?  Performance?

Author: Thanks, has been fixed.

>    - Potential more CPU overhead in the Service VM if using ``virtio-net``.

Author: Thanks, has been fixed.

Inter-VM shared memory communication (ivshmem)
**********************************************

Inter-VM shared memory communication is based on a shared memory mechanism

> to transfer the data between VMs. The ACRN device model or hypervisor emulates

Author: Thanks, has been fixed.

a virtual PCI device (called an ``ivshmem device``) to expose this shared memory's
base address and size. (Refer to :ref:`ivshmem-hld` and :ref:`enable_ivshmem` for the
background introductions).

:numref:`Inter-VM shared memory communication` shows the Inter-VM shared memory communication overview:

.. figure:: images/Inter-VM_shared_memory_communication_overview.png
   :align: center
   :name: Inter-VM shared memory communication

   Inter-VM shared memory communication

- Pros:
   - Shared memory is exposed to VMs via PCI MMIO Bar and is mapped and accessed directly.

>    - The data path is short.

Author: Thanks, has been fixed.

>    - No Service VM emulation is involved if using hv-land ivshmem.

Author: Thanks, has been fixed.

- Cons:

>    - Applications talk to the device directly (via `UIO <https://doc.dpdk.org/guides/linux_gsg/linux_drivers.html#uio>`_

Reviewer 2: Why is this a con?  Seems like it will get you fast performance.  Or is the con really your second point below?

Author: Thanks, has been fixed.

     on Linux and `Ivshmem driver <https://github.com/virtio-win/kvm-guest-drivers-windows/tree/master/ivshmem/>`_
     on Windows).
   - Applications need to implement protocols such as a handshake, data transfer, and data
     integrity.

How to implement an Ivshmem application on ACRN
***********************************************

As Ivshmem is a PCI device that shares memory inside a virtual machine, the application

> running on the VM can access it as a standard PCI device. Besides that, the application

Reviewer 2: ```suggestion
running on the VM can access it as a standard PCI device. The application
```

Author: Thanks, has been fixed.

.. _inter-vm_communication:

ACRN Inter-VM Communication
##############################################

ACRN supports three kinds of Inter-VM communication – Inter-VM vUART,

> Inter-VM network communication, and Inter-VM shared memory device (ivshmem).

Author: Thanks, has been fixed.

Each communication method has its pros and cons, as described below.

Inter-VM vUART
**************

Inter-VM vUART communication is based on the vUART implementation. It is

> used to transfer data between two VMs at low speed(< 120kbps). (Refer to :ref:`vuart_virtualization`

Reviewer: ```suggestion
used to transfer data between two VMs at low speed (< 120kbps). (Refer to :ref:`vuart_virtualization`
```

Author: Thanks, has been fixed.

and :ref:`vuart_config`).

:numref:`Inter-VM vUART communication` shows the Inter-VM vUART communication overview:

> .. figure:: images/Inter-VM_vUART_communocation_overview.png

Reviewer: Nitpick: can we update the filename to correct the typo in it?

Author: Thanks, has been fixed.

   :align: center
   :name: Inter-VM vUART communication

   Inter-VM vUART communication

- Pros:
   - POSIX APIs.

>    - UART driver is available in most VM, development-friendly(POSIX can be called directly

Reviewer: I think I understand your statement ("development-friendly") better now... I like it but I'd like it to be rephrased somehow, along the lines of: "Easily used programmatically by using standard POSIX APIs")

Author: Thanks, has been fixed.

>    - UART driver is available in most VMs.

Reviewer: I think it would more accurate that the "UART drivers are available in most Operating Systems", wouldn't it?

Author: yes, Thanks, has been fixed.

- Cons:
   - The communication rate is low.
   - Need to configure communication vUART in hypervisor scenario.

Inter-VM network communication
******************************

Inter-VM network communication is based on the network stack. ACRN supports
both pass-through NICs to VM and Virtio-Net solutions. (Refer to :ref:`virtio-net`
background introductions of ACRN Virtio-Net Architecture and Design). 

:numref:`Inter-VM network communication` shows the Inter-VM network communication overview:

.. figure:: images/Inter-VM_network_communication_overview.png
   :align: center
   :name: Inter-VM network communication

   Inter-VM network communication

- Pros:

>    - Socket-based APIs; development friendly.

Reviewer: Same comment here, it would be good to rephrase somehow to indicate this is used via the classical socket API (and hence is a well know mechanism for many developers).

@dbkinder @amyreye , could you perhaps suggest some text to that effect?

- Cons:
   - Multiple layers are involved across the data path.
   - Potential more CPU overhead in the Service VM if using ``virtio-net``.

Inter-VM shared memory communication (ivshmem)
**********************************************

Inter-VM shared memory communication is based on a shared memory mechanism
to transfer the data between VMs. The ACRN device model or hypervisor emulates

> a virtual PCI device (called an ivshmem device) to expose this shared memory's

Reviewer: ```suggestion
a virtual PCI device (called an ``ivshmem`` device) to expose this shared memory's
```

Author: Thanks, has been fixed.

base address and size. (Refer to :ref:`ivshmem-hld` and :ref:`enable_ivshmem` for the
background introductions).

:numref:`Inter-VM shared memory communication` shows the Inter-VM shared memory communication overview:

.. figure:: images/Inter-VM_shared_memory_communication_overview.png
   :align: center
   :name: Inter-VM shared memory communication

   Inter-VM shared memory communication

- Pros:
   - Shared memory is exposed to VMs via PCI MMIO Bar and is mapped and accessed directly.
   - The data path is short.
   - No Service VM emulation is involved if using hv-land ivshmem.

- Cons:

>    - Applications talk to the device directly (via UIO on Linux and Ivshmem driver on Windows).

Reviewer: Can we add links to UIO on Linux and the Windows driver?

Author: Thanks, has been added.

   - Applications need to implement protocols such as a handshake, data transfer, and data
     integrity.

How to implement an Ivshmem application on ACRN
***********************************************

As Ivshmem is a PCI device that shares memory inside a virtual machine, the application
running on the VM can access it as a standard PCI device. Besides that, the application
needs to design a data transfer notification mechanism between the VMs.

**The following reference code are used for Inter-VM communication using ivshmem device**

> - Initial Device:

Author: Thanks, has been fixed.

  Enable and bind to UIO PCI driver

>     .. code-block:: none

Author: Thanks, has been fixed.

       modprobe uio
       modprobe uio_pci_generic
       echo "1af4 1110" > /sys/bus/pci/drivers/uio_pci_generic/new_id

>     .. note:: "1af4" is the Vendor ID and "1110" is the Device ID of the ivshmem device.

Author: Thanks, has been fixed.

    .. note:: For Linux-based User VMs, we recommend using the standard UIO and UIO_PCI_GENERIC

>      drivers through the device node (for example, ``/dev/uioX``)

Author: Thanks, has been fixed.

> - Uio irq data Struct

Reviewer 2: ```suggestion
- UIO IRQ data struct
```

Author: Thanks, has been fixed.

>      .. code-block:: none

Reviewer 2: As before, this code block directive should be indented to the "U"  above, and the contents of the code block indented with the "c" in code-block.  Also, since this code block is C code, use the language "c" instead of "none" for proper highlighting.

Author: Thanks, has been fixed.

            for (i = 0; i < IVSH_MAX_IRQ_NUM; i++) {
                p_ivsh_dev_ctx->epfds_irq[i] = -1;
                p_ivsh_dev_ctx->irq_data[i].fd = -1;
            }
            p_ivsh_dev_ctx->opened = false;
            return 0;
        }

- Get ivshmem device shared memory size

     .. code-block:: none
        :emphasize-lines: 10

        uint32_t ivsh_get_shmem_size(long uio_nr)
        {
            char config_node[PATH_MAX] = {0};
            uint32_t shm_size;
            uint64_t tmp;
            int cfg_fd;

            sprintf(config_node, "/sys/class/uio/uio%ld/device/config", uio_nr);
            cfg_fd = open(config_node, O_RDWR);
            /*Get the size of BAR2(offset in Configuration Space is 0x18) of uio_nr device*/
            pread(cfg_fd, &tmp, 8, 0x18);
            shm_size= ~0U;
            pwrite(cfg_fd ,&shm_size, 8, 0x18);
            pread(cfg_fd, &shm_size, 8, 0x18);
            pwrite(cfg_fd ,&tmp, 8, 0x18);
            shm_size &= (~0xfUL);
            shm_size = (shm_size & ~(shm_size - 1));
            close(cfg_fd);

            return shm_size;
        }

- Open Ivshmem Device:

     .. code-block:: none

>         :emphasize-lines: 1,6,12,18,22

Reviewer 2: I'm curious why you're emphasizing some of the comments

Author: Thanks, has deleted these emphasize-lines.

                    p_ivsh_dev_ctx->epfds_irq[i] = -1;
                end
            close(p_ivsh_dev_ctx->uio_dev_fd);
            p_ivsh_dev_ctx->uio_dev_fd = -1;
        }

**The following reference code are used for Inter-VM communication based on Doorbell mode:**

- Trigger ivshmem doorbell

     .. code-block:: none

        void ivsh_trigger_doorbell(struct ivsh_dev_context *p_ivsh_dev_ctx, uint16_t peer_id, uint16_t vector_id)
        {
            p_ivsh_dev_ctx->p_reg[IVSH_REG_DOORBELL >> 2] = (peer_id << 16) | vector_id;
        }

- Wait ivshmem device irq

     .. code-block:: none

        static inline int ivsh_wait_irq(struct ivsh_dev_context *p_ivsh_dev_ctx, unsigned int idx)
        {
            struct epoll_event ev = {0};
            struct uio_irq_data *irq_data = NULL;
            eventfd_t val;
            int n;
            while (1) {
                n = epoll_wait(p_ivsh_dev_ctx->epfds_irq[idx], &ev, 1, -1);
                if (n == 1) {
                    irq_data = ev.data.ptr;
                    eventfd_read(irq_data->fd, &val);
                    break;
                }
            }
        }

> **Data Transfer State-Machine**

Reviewer 2: This should probably be a subheading:
```suggestion
Data Transfer State-Machine
===========================
```

Author: Thanks, has been fixed.

>   A state machine is introduced as a communication mechanism between the two VMs,

Reviewer 2: Remove the leading extra space indent on all these lines

  which use the same ivshmem PCI device for the data transfer.

  It includes three states – RESET, READY, and INIT. RESET state is the initial state
  after ivshmem device is initialized.

>     - When both VM states are in the RESET, Sender VM prepares the sending data

Reviewer 2: The bullet list should not be indented relative to the paragraph above.

Author: Thanks, has been fixed.

      and then sets its state to INIT, and Receiver VM prepares receiving buffer
      and sets its state to INIT.

    - When both VM's state is in the INIT, Sender VM sets its state in the READY
      after sending all of the data, and Receiver VM sets its state in the READY
      after receiving all of the transfer data.

    - Then both VMs change their status from READY to INIT for starting the next
      round of data transfer.

  :numref:`Inter-VM ivshmem data transfer state machine` shows the state machine relationship:

>     .. figure:: images/Inter-VM_data_transfer_state_machine.png

Reviewer 2: Watch the indenting of all this figure directive and the sentence above.  If it's not part of the bullet list, it should not be indented at all.

Author: Thanks, has been fixed.

       :align: center
       :name: Inter-VM ivshmem data transfer state machine

       Inter-VM ivshmem data transfer state machine

  :numref:`Inter-VM ivshmem handshake communication` shows the handshake communication between two machines:

    .. figure:: images/Inter-VM_handshake_communication_two_machine.png
       :align: center
       :name: Inter-VM ivshmem handshake communication

       Inter-VM ivshmem handshake communication

> **The reference Sender and Receiver Sample Code based Doorbell mode:**

Reviewer 2: This should probably be a subheading too...
```suggestion
Reference Sender and Receiver Sample Code Based Doorbell Mode
=============================================================
```

Author: Thanks, has been fixed.

>      .. code-block:: none

Reviewer 2: This code-block should not be indented.

Author: Thanks, has been fixed.

            ivsh_test_tx_send(&tx_ctx);
            ivsh_trigger_doorbell(p_ivsh_dev_ctx, p_ivsh_ctrl_ctx->peer_id, IVSH_TEST_VECTOR_ID);
            ivsh_wait_irq(p_ivsh_dev_ctx, IVSH_TEST_VECTOR_ID);

            ivsh_test_tx_deinit(&tx_ctx);
        }

        void ivsh_test_receiver(struct ivsh_dev_context *p_ivsh_dev_ctx, struct ivsh_ctrl_context *p_ivsh_ctrl_ctx)
        {
            struct ivsh_test_rx_context rx_ctx;
            volatile struct pay_load_header *p_hdr;

            /* Initialize the receiver related data */
            ivsh_test_rx_init(&rx_ctx, p_ivsh_dev_ctx, p_ivsh_ctrl_ctx);
            p_hdr = rx_ctx.p_hdr;
            while (!is_p0_reset(p_hdr))
                usleep(10000);
            set_p1_status(p_hdr, SHMEM_STATUS_RESET);

            while (!is_p0_initialized(p_hdr))
                usleep(100);
            set_p1_status(p_hdr, SHMEM_STATUS_INIT);
            set_p0_status(p_hdr, SHMEM_STATUS_READY);

            /* waiting for p0 write done */
            ivsh_wait_irq(p_ivsh_dev_ctx, IVSH_TEST_VECTOR_ID);
            ivsh_test_rx_recv(&rx_ctx);
            usleep(100);

            ivsh_test_rx_deinit(&rx_ctx);
        }

> **The reference Sender and Receiver Sample Code based Polling mode:**

Reviewer 2: Also better as a subheading
```suggestion
Reference Sender and Receiver Sample Code Based Polling Mode
============================================================
```

Author: Thanks, has been fixed.

>      .. code-block:: none

Reviewer 2: Same here.  Remove the code block indentation.

Author: Thanks, has been fixed.
